teuthology/suite/run.py

```diff
@@ -196,15 +196,58 @@ def choose_ceph_version(self, ceph_hash):
             log.info('skipping ceph package verification')
 
     def choose_teuthology_branch(self):
+        """Select teuthology branch, check if it is present in repo and
+        return the branch name value.
+
+        The branch name value is determined in the following order:
+
+        Use ``--teuthology-branch`` argument value if supplied.
+
+        Use ``TEUTH_BRANCH`` environment variable value if declared.
+
+        If yaml file ``qa/.teuthology`` can be found in suite repo
+        supplied with ``--suite-repo`` or ``--suite-dir`` and
+        has ``teuthology_branch`` record then use its value.
+
+        Use ``teuthology_branch`` value if it is set in the one
+        of the teuthology config files ``$HOME/teuthology.yaml``
+        or ``/etc/teuthology.yaml`` correspondingly.
+
+        Use ``--suite-branch`` argument value if supplied.
+
+        Use ``--ceph`` argument value.
+
+        Use ``master`` if the branch supplied with ``--suite-branch``
+        or ``--ceph`` is missing in the teuthology repo.
+
+        Generate exception if the branch is not present in the repo.
+        """
         teuthology_branch = self.args.teuthology_branch
-        if teuthology_branch and teuthology_branch != 'master':
-            if not util.git_branch_exists('teuthology', teuthology_branch):
-                exc = BranchNotFoundError(teuthology_branch, 'teuthology.git')
-                util.schedule_fail(message=str(exc), name=self.name)
-        elif not teuthology_branch:
+        if teuthology_branch:
+            if  teuthology_branch != 'master':
+                if not util.git_branch_exists('teuthology', teuthology_branch):
+                    exc = BranchNotFoundError(teuthology_branch, 'teuthology.git')
+                    util.schedule_fail(message=str(exc), name=self.name)
+        elif 'TEUTH_BRANCH' in os.environ:
+            teuthology_branch = os.environ.get('TEUTH_BRANCH')
+        else:
             # Decide what branch of teuthology to use
-            if util.git_branch_exists('teuthology', self.args.ceph_branch):
-                teuthology_branch = self.args.ceph_branch
+            check_branch = None
+            branch_config = self.suite_repo_path + '/qa/.teuthology'
+            if os.path.exists(branch_config):
+                log.debug('Found teuthology config in suite branch %s', branch_config)
+                with open(branch_config) as f:
+                    y = yaml.safe_load(f)
+                    check_branch = y.get('teuthology_branch', None)
+                    if check_branch:
+                        log.debug('Teuthology branch is overridden with %s',
+                                  check_branch)
+            if not check_branch:
+                check_branch = config.get('teuthology_branch', None)
+            if not check_branch:
+                check_branch = self.args.suite_branch or self.args.ceph_branch
```

Comment:
I agree with Ilya about dropping suite_branch and ceph_branch fallbacks - they're no longer necessary

Reply:
dropped